[SW-2216] ScalaInt, DataFrames, H2oFrames and RDDS endpoints must be handled on Spark driver side as we require both Spark and h2o features

jakubhava opened the pull request:
The goal of this change is to ensure that endpoints for RDDs, DataFrames, ScalaInterpreter and H2OFrames remain functional after we move towards client-less approach. In fact, these endpoints haven't been working in client-less approach already.

We needed to ensure that all the requests are passed to proxy, but requests for these endpoints must be handled on the driver as we require both H2O and Spark functionality.

jakubhava commented:
Still need to make sure that async mode of Scala cells works by updating a little bit ScalaInterpreterServler and writing tests, but at this stage I think this change is ready for review

mn-mikke commented on core/src/main/scala/ai/h2o/sparkling/backend/api/DELETERequestBase.scala:
```diff
@@ -0,0 +1,37 @@
+
+  def handleDeleteRequest(request: HttpServletRequest): Any
+
+  override def doDelete(req: HttpServletRequest, resp: HttpServletResponse): Unit = {
```
The body of `doDelete` and [`doGet`](https://github.com/h2oai/sparkling-water/pull/2093/files#diff-ff17f4222dc68216fd015c4db186a978R27) seems to be identical.

mn-mikke commented on core/src/main/scala/ai/h2o/sparkling/backend/api/rdds/RDDsServlet.scala:
```diff
@@ -0,0 +1,113 @@
+  override def handleGetRequest(request: HttpServletRequest): Any = {
+    request.getRequestURI match {
+      case "/3/RDDs" =>
+        list()
```
nit: can be on one line?

jakubhava replied:
👍

mn-mikke commented on core/src/main/scala/ai/h2o/sparkling/backend/api/rdds/RDDsServlet.scala:
```diff
@@ -0,0 +1,113 @@
+
+  override def handleGetRequest(request: HttpServletRequest): Any = {
+    request.getRequestURI match {
+      case "/3/RDDs" =>
```
Will this work with a context path?

jakubhava replied:
Good point, will add test as well

jakubhava replied:
Good catch, did not work, but should be fine now + tests for changed endpoints now run with context path

mn-mikke commented on core/src/main/scala/ai/h2o/sparkling/backend/api/scalainterpreter/ScalaInterpreterServlet.scala:
```diff
@@ -0,0 +1,188 @@
+      jobCount.incrementAndGet()
+      while (conf.maxParallelScalaCellJobs != -1 && jobCount
+               .intValue() > conf.maxParallelScalaCellJobs) {
+        Thread.sleep(1000)
```
Why do we need `Thread.sleep` here?

jakubhava replied:
There are two scala cell execution modes -> synchronous and asynchronous. In synchronous we start the computation and get result right away. In asynchronous, the scala cell starts computing in a form on H2O Job and once H2O has result, Flow UI is able to interpret it. 

There was a customer requesting this feature some time ago. The sleep here is to ensure that we don't start more than `conf.maxParallelScalaCellJobs` parallel jobs in asynchronous mode

jakubhava replied:
Note: the sleep should no longer be needed, I have rewrote this piece of code using using executor service

mn-mikke commented on core/src/main/scala/ai/h2o/sparkling/backend/api/scalainterpreter/ScalaInterpreterServlet.scala:
```diff
@@ -0,0 +1,188 @@
+  def interpret(sessionId: Int, code: String): ScalaCode = {
+    this.synchronized {
+      jobCount.incrementAndGet()
+      while (conf.maxParallelScalaCellJobs != -1 && jobCount
```
nit: `jobCount` to the next line

jakubhava replied:
👍 I rewrote the code a bit to get rid of the sleep bellow using executor pool service

mn-mikke commented on extensions/src/main/scala/ai/h2o/sparkling/extensions/rest/api/Paths.scala:
```diff
@@ -24,4 +24,8 @@ object Paths {
   val FINALIZE_FRAME: String = "/3/FinalizeFrame"
   val UPLOAD_PLAN: String = "/3/UploadPlan"
   val LOG_LEVEL: String = "/3/LogLevel"
+  val SW_INTERNAL: String = "/3/sw_internal/*"
```
nit: Can we use the same naming conventions for path?

jakubhava replied:
👍

mn-mikke commented on extensions/src/main/scala/ai/h2o/sparkling/extensions/rest/api/InternalUtilsServlet.scala:
```diff
@@ -0,0 +1,86 @@
+          job.start(new H2OCountedCompleter() {
+            override def compute2(): Unit = {
+              while (jobs.contains(jobV3.key.name) && jobs(jobV3.key.name)) {
+                Thread.sleep(100)
```
Why do we need `Thread.sleep(100)`?

jakubhava replied:
I have tried to document this in the code:

``
/**
 * In Asynchronous mode of Scala cell execution H2O Flow expects that the cell returns H2O Job right away and then
 * when clicking on the job link, the user will either get informed that the job is still running or the result will
 * be displayed.
 *
 * We need to perform the computation on Spark Driver to have access to both H2O and Spark and since H2O expects the H2O
 * Job object, we are submitting dummy H2O job which is alive for the duration of the computation on the Spark side.
 *
 * That is the reason for submitDummyWrapperH2OJob and stopDummyWrapperH2OJob methods.
 */
``

That is also the reason for the sleep. We need to wait until thread on Spark driver finishes before we can mark this dummy job as stopped.

mn-mikke commented on extensions/src/main/scala/ai/h2o/sparkling/extensions/rest/api/InternalUtilsServlet.scala:
```diff
@@ -0,0 +1,86 @@
+
+    processRequest(request, response) {
+      val obj = request.getRequestURI match {
+        case s if s.startsWith("/3/sw_internal/start/") =>
```
nit: Can we break this branch to more methods to improve readability?

jakubhava replied:
👍

jakubhava commented on core/src/main/scala/ai/h2o/sparkling/backend/api/dataframes/DataFrameInfo.scala:
```diff
@@ -0,0 +1,43 @@
+
+  private[dataframes] case class DataFrameInfoParameters(dataFrameId: String) {
+    def validate(): Unit = {
+      if (!SparkSessionUtils.active.sqlContext.tableNames().toList.contains(dataFrameId)) {
```
👍

jakubhava commented on core/src/main/scala/ai/h2o/sparkling/backend/api/ServletBase.scala:
```diff
@@ -16,13 +16,10 @@
-  override def toString: String = name
+private[api] trait ServletBase extends HttpServlet {
+  protected def toScalaRegex(pathSpec: String): String = {
+    pathSpec.replaceAll("\\*", "\\.*")
```
The intent here is to replace all occurrences of `*` by `.*` as we are trying to create Scala regex from Jetty regex path spec. I will double verify this is correct

jakubhava replied:
Yup, just verified manually ( plus all tests pass)

val s = "/3/frames/*/h2oframes\\"
leads to `/3/frames/*/h2oframes\`

jakubhava commented on core/src/main/scala/ai/h2o/sparkling/backend/api/dataframes/DataFrameInfo.scala:
```diff
@@ -0,0 +1,43 @@
+object DataFrameInfo extends ParameterBase {
+
+  private[dataframes] case class DataFrameInfoParameters(dataFrameId: String) {
+    def validate(): Unit = {
```
👍

jakubhava commented on core/src/main/scala/ai/h2o/sparkling/backend/api/dataframes/DataFrameToH2OFrame.scala:
```diff
@@ -0,0 +1,44 @@
+
+  object DataFrameToH2OFrameParameters {
+    private[dataframes] def parse(request: HttpServletRequest): DataFrameToH2OFrameParameters = {
+      val dataFrameId = request.getRequestURI.split("/")(3)
```
Good question! I will write test for it

jakubhava replied:
Good catch, did not work, but should be fine now + tests for changed endpoints now run with context path

jakubhava commented:
@mn-mikke I hope I have addressed all the comments. Can you please have a second go? I have also tried to comment the tricky parts in the code. Hopefully it will make better sense now, otherwise please keep the questions coming!

mn-mikke commented on core/src/main/scala/ai/h2o/sparkling/backend/api/dataframes/DataFramesRestApi.scala:
```diff
@@ -0,0 +1,45 @@
+  protected def listDataFrames(): DataFrames = {
+    val hc = H2OContext.ensure()
+    val endpoint = new URI(hc.flowURL())
+    query[DataFrames](endpoint, "/3/dataframes", hc.getConf)
```
nit: Doesn't existing path start with a capital letter `/3/Dataframes`? I'm not sure, just asking..

jakubhava replied:
Good questions! It is lowercase ( inconsistent, but would like the endpoints to remain the same as part of this change)

jakubhava commented:
Thank you a lot for review!
